Add a load workflows action to the redux classify store

eatyourgreens opened the pull request:
Staging branch URL: https://pr-5124.pfe-preview.zooniverse.org

Add a `loadWorkflows` action creator to the classifier. Given a workflow ID, locale and optional preferences, it clears the current workflow, empties the subject queue, loads the workflow translations, loads the workflow and updates preferences to store the workflow ID. If the API request errors, the stored workflow is set to null and the preference changes are rolled back.

Classifier state is reset before loading a new workflow, which fixes #5149.

# Required Manual Testing

- [x] Does the non-logged in home page render correctly?
- [x] Does the logged in home page render correctly?
- [x] Does the projects page render correctly?
- [x] Can you load project home pages?
- [x] Can you load the classification page?
- [x] Can you submit a classification?
- [ ] Does talk load correctly?
- [ ] Can you post a talk comment?

# Review Checklist

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [x] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [ ] If the component is in coffeescript, is it converted to ES6? Is it free of eslint errors? Is the conversion its own commit?
- [x] Are the tests passing locally and on Travis?

# Optional

- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you [resized and compressed](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/image-optimization) any image you've added?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?
- [ ] Have you followed the [Springer guidelines for commit messages](https://github.com/springernature/frontend-playbook/blob/master/git/git.md#commit-messages)?

eatyourgreens commented:
`docker-compose run dev npm run test-one app/redux/ducks/classify.spec.js` is useful when trying out the new tests.

eatyourgreens commented:
Rebased off #5137, which needs to be merged first.

eatyourgreens commented:
I've changed the base branch to `translations`, to make the changes easier to see.

eatyourgreens commented:
This should be ready, once Travis and Jenkins have finished. The base branch will need to be changed from `translations` to `master` after #5137 is merged.

eatyourgreens commented on app/redux/ducks/classify.spec.js:
```diff
@@ -454,4 +456,96 @@ describe('Classifier actions', function () {
+    });
+
+    it('should load the workflow translation', function () {
+      const loadTranslations = fakeDispatch.returnValues[3];
```
Is there a better way of spying on this call than looking for the fourth call to `dispatch()`? Sinon won't spy on pure functions, so `translations.load` can't be stubbed.

eatyourgreens commented:
Rebased now that #5137 is merged.

shaunanoordin commented:
## PR Comments

To clarify the purpose of this PR:
1. the intent of this PR is to consolidate the the _"fetch my workflow"_ logic, correct?
2. the _"fetch my workflow"_ logic can best be explained as _"given workflow ID+preferences+language, be sure to fetch BOTH translations and the workflow before proceeding (flagging that a workflow has been successfully fetched),"_ ?
3. the ultimate goal is to make sure workflow resources and translation resources are _resolved_ at the same time, instead of one racing the other due to network unpredictability?

Please correct me if I'm wrong on any of the above.

As far as I can tell, the new `loadWorkflow()` duck works properly in the following scenario:
- at a Project home page with multiple workflow buttons to choose from, clicking on any workflow button will trigger the `loadWorkflow()` logic.
  ![screen shot 2018-12-14 at 17 21 54](https://user-images.githubusercontent.com/13952701/50018090-0bb7c300-ffc6-11e8-89e3-a7bab2635f37.png)
  _Example project with multiple workflows. Clicking on "Easy Arabic" triggers `loadWorkflow()`_

However, there are three scenarios where the `loadWorkflow()` doesn't seem to trigger, and instead some other function/method is used to successfully fetch workflows:
- when a user clicks on the default "Get Started" link that simply directs the user to the `/classify` page.
- when a user accesses the `/classify` page directly, e.g. via bookmark
- when a user accesses a specific workflow via the `/classify` page directly, e.g. `https://localhost:3735/projects/darkeshard/transformers/classify?reload=0&workflow=3262`

Also, there's one scenario where I'm **not sure if** the new consolidated `loadWorkflow()` logic _is supposed_ to trigger:
- when a user changes the language of the project
  ![screen shot 2018-12-14 at 17 22 01](https://user-images.githubusercontent.com/13952701/50018155-43266f80-ffc6-11e8-8e17-4d5cd5599429.png)
  _In this example, Variable Star Zoo has two languages (English, Espanol) selectable in the top-right language dropdown. Changing this dropdown - whether on the project home page, or the Classify page - does not trigger `loadWorkflow()`_
 
Is `loadWorkflow()` supposed to trigger in those scenarios, or are those scenarios working as expected?

shaunanoordin commented:
Wait, goddanggit, I just saw a funky fresh new push at https://github.com/zooniverse/Panoptes-Front-End/pull/5124#event-2027395917, right before I posted that PR Comments post. Let me run through the tests again.

shaunanoordin commented:
OK, I've just refreshed my `load-workflows` branch and my observations from the PR Comments post remain the same. I'd just like to confirm that `loadTranslations()` is only supposed to work in the specific scenario where users chooses a workflow by clicking on a Workflow button on the Project home page, and not in any other scenario.

EDIT: just realise I need to test what happens when a Project has both selectable Workflow buttons AND the default Get Started button on the home page. Eh, that's a Monday problem.

srallen commented:
> EDIT: just realise I need to test what happens when a Project has both selectable Workflow buttons AND the default Get Started button on the home page. Eh, that's a Monday problem.

There shouldn't be any projects like this. If there are, then that's a UI bug. It should be an either/or UI state.

eatyourgreens commented:
It would be good to get this action into the `WorkflowSelection` component too, if that's possible. That would handle the cases where @shaunanoordin noticed that it isn't being called.

It shouldn't be called on language change. In that case, we load a new translation but the workflow is already in the store, and preferences don't need updating.

At some point, I might update user prefs on language change so we remember which language you prefer, but that's a separate issue.

eatyourgreens commented on app/redux/ducks/classify.spec.js:
```diff
@@ -454,4 +456,106 @@ describe('Classifier actions', function () {
+    });
+
+    beforeEach(function () {
+      state = initialState;
```
`Object.assign` should be used here so that we don’t mutate `initialState`.

eatyourgreens commented on app/redux/ducks/classify.spec.js:
```diff
@@ -454,4 +456,106 @@ describe('Classifier actions', function () {
+      if(typeof action === 'function') {
+        action = action(fakeDispatch);
+      }
+      storeState = reducer(state, action);
```
`state` is always the initial state here. The dispatcher should be applying the action to `storeState`.

eatyourgreens commented:
I've updated this so that loading a new workflow resets the classify store to its initial state.

eatyourgreens commented:
@shaunanoordin this PR also updates the workflow assignment dialog, which I think is only used on Gravity Spy. It's essentially a variant of the home workflow buttons, which promotes you to a new workflow when you confirm the dialog.

eatyourgreens commented on app/redux/ducks/classify.js:
```diff
@@ -346,6 +363,42 @@ export function saveAnnotations(annotations) {
+    });
+    const awaitTranslation = dispatch(translations.load('workflow', workflowId, locale));
+    return Promise.all([awaitWorkflow(workflowId), awaitTranslation])
+    .then(([workflow]) => {
```
Is that something that could happen when this action is called from  the workflow assignment dialog? I’m assuming you can’t try to load an inactive workflow from the project home.

eatyourgreens replied:
Looking at the API call, it should error with a 404 status if there’s no workflow, but I haven’t tested that to make sure.

srallen replied:
Yeah, it's possible from workflow assignment since whether or not those ids in the UPP are active isn't checked until we do the request (this eventually could be refactored to avoid the API request). I thought the activeness param would return an empty array and 200 if the workflow requested is no longer active, though I could be misremembering, and this might be the behavior for workflows you do not have permission to (workflows in development on a non-live project?). Regardless, I think there should be a existence check here before saving UPP and dispatching to the Redux store.

srallen replied:
I looked it over and lines 115-136 in the workflow selection component are what checks for the existence and has slightly different handling than the catch. I think I was correct in thinking that these are inactive workflows being requested for. Could be from ids stored in UPP, and possibly other cases. There were cases of inactive workflows still being project defaults before the lab got fixed too.

eatyourgreens replied:
With the current code, if you are promoted to an inactive workflow, the classifier will simply load the workflow by ID without giving any indication that it's inactive. However, you won't be able to select the new workflow from the project home page.

srallen replied:
AH, I see what you mean. I think that checking against the serialized active workflow ids array was added a bit later for all of the workflow ids. 

I believe I added the existence safety check prior to the serialization of the active workflow links. What I'm suggesting, though, is that we still do an existence check in the response and add handling if the response array is empty because it is possible as I've run into situations that I didn't think of and workflow selection breaking because of it. I can't see that happening reading it over again, but I've been burned in the past couple of years from some buggy situations.

eatyourgreens replied:
Ok, but i think that an empty response from `GET /workflows/:id` would be a bug in Panoptes. Requesting a specific resource should return something or an error.

eatyourgreens replied:
It occurred to me that `GET /workflows/?workflow_id={:id}&active=true` would return an empty array for inactive workflows, since searches return arrays, but this code doesn't use that route. Neither does workflow selection. We removed that from workflow selection because the code should request a workflow directly when the ID is known, rather than running a search across the workflows table.

This code and workflow selection use an endpoint that returns either a single resource or a 404 error, and the 404 is caught by the catch block.

There's also a bug in workflow selection where API errors run the error handling code twice. I'll open an issue for that.

eatyourgreens replied:
I've updated the `then` block to raise a 404 error, and clear stored preferences, if the workflow doesn't exist. If that case does happen, we really should look at why it's happening and fix Panoptes.

srallen replied:
Good point on the search param version of the request. Looking back through commit history, that's what we used to use, so it made sense then to check whether or not the response array is empty.

eatyourgreens replied:
From memory, the original code caused performance problems when Panoptes was under heavy load. 

It’s returning an array now because `Promise.all` returns `[workflow, translation]`. I debated adding the translation to make the code easier to understand, but left it out because the linter complains about declaring variables that aren’t used.

eatyourgreens commented on app/pages/project/home/home-workflow-button.jsx:
```diff
@@ -16,21 +15,17 @@ class ProjectHomeWorkflowButton extends React.Component {
     this.handleWorkflowSelection = this.handleWorkflowSelection.bind(this);
   }
 
+  componentDidUpdate() {
```
That’s a good idea. My only concern is that it removes the option to right click and open in a new tab.

eatyourgreens replied:
`HomeWorkflowButton` is now a button, which makes me a lot happier with the component name.

eatyourgreens commented:
That error sounds as if `fetchSubjects` is being called with a null workflow. That shouldn’t be happening.

eatyourgreens commented:
I've rebased this to the latest master. I've also added a check for the workflow before refilling the subject queue and updated the load workflows action to clear the project set workflow after load. My understanding is that once the load request either resolves or rejects, the ID set by the project owner will either be stored in user preferences, or invalid, so I think it should be safe to clear it.

eatyourgreens commented:
I've updated the workflow assignment prompt code to check for active workflow IDs, but I'm not 100% sure if this is the right place in the code to be checking for active workflows.

eatyourgreens commented on app/redux/ducks/classify.js:
```diff
@@ -346,6 +363,48 @@ export function saveAnnotations(annotations) {
+    })
+    .then((action) => {
+      if (preferences && preferences.settings && preferences.settings.workflow_id === workflowID) {
+        preferences.update({ 'settings.workflow_id' : undefined });
```
I think preferences should be saved here, to catch all changes.

srallen commented:
> I've updated the workflow assignment prompt code to check for active workflow IDs, but I'm not 100% sure if this is the right place in the code to be checking for active workflows.

I think validating the id should remain in the workflow selection code. In the future, this could be handled in a store for the UPP when the UPP loads, but what I don't want is to have validations scattered throughout the code base like we have with the API requests right now. I'd like to see a standard API for workflow selection as actions for a store and the components just call the action for that when necessary. 

> Also, I'm a little confused that shouldWorkflowAssignmentPrompt uses this.props and nextProps. Would anything break if it used this.props consistently throughout?

Yes, it would break if that was the only thing that changed. It's comparing the incoming next props for `settings.workflow_id` to the current prop for `preferences.selected_workflow`. If anything does change, it would be `this.props` to `nextProps` on line 156 since it is unlikely that both UPP attributes would be changing. Alternatively, refactor that method call from `componentWillReceiveProps` to `componentDidUpdate` and use `prevProps` and `this.props`. I think that's out of scope for this PR though.

eatyourgreens commented:
`componentWillReceiveProps` is deprecated so moving to `componentDidUpdate` is what I'd prefer.

eatyourgreens commented:
I missed a couple of things yesterday when converting the workflow links to buttons, but this should be ready now.

eatyourgreens commented:
I've also updated workflow selection to use `Promise.all` to fetch the workflow and its translation, to avoid conditions where we update workflow state while the translation is still loading.

eatyourgreens commented on app/pages/project/classify.jsx:
```diff
@@ -148,15 +150,22 @@ export class ProjectClassifyPage extends React.Component {
         nextProps.preferences.settings &&
         nextProps.preferences.settings.workflow_id;
-      const currentWorkflowID = this.props.preferences && this.props.preferences.preferences.selected_workflow;
+      const currentWorkflowID = preferences && preferences.preferences.selected_workflow;
```
Current workflow ID will actually be `this.props.workflow.id`, which should be the same as user prefs but we can't assume that.

eatyourgreens commented:
I thought about converting `componentWillReceiveProps` to `componentDidUpdate` on the classify page, but this PR has grown a lot and my impression is it might already have too many changes to comfortably review?

srallen commented:
@eatyourgreens I'm a little anxious about merging this before the holiday break as it's a bit more to review now. Could the changes to the workflow assignment be cherry-picked into its own branch and PR? Then this can be scoped to just adding the workflow action and refactoring the home page buttons?

srallen commented:
Additionally, the adding of a reset action for the classification could be cherry-picked out so we have a PR just to fix the gold standard setting persisting between projects.

eatyourgreens commented:
I'm not sure. I guess the reset action would replace current code that calls `setWorkflow(null)`?

The changes to workflow assignment go hand in hand with the changes here, so I'm not sure how to cleanly separate them out. At least, workflow assignment should be calling `loadWorkflow` and should only happen if the workflow is valid. I don't think the popup should appear if clicking it resets the workflow to null?

srallen commented:
Ok, I'll look this over and see if I can spot a way this can be broken apart. If not, then I'll do a full review again as is.

eatyourgreens commented:
I think the changes to `shouldWorkflowAssignmentPrompt` could be moved into a small PR of their own. They validate the workflow ID so that you aren't prompted to load a new workflow if that workflow doesn't exist.

eatyourgreens commented on app/pages/project/classify.jsx:
```diff
@@ -119,11 +119,13 @@ export class ProjectClassifyPage extends React.Component {
-    .catch((error) => {
-      this.setState({ rejected: { classification: error }});
-    });
+    if (workflow) {
```
Should this if wrap `maybePromptWorkflowAssignment` too?

srallen replied:
Yeah probably, but I think more about that in the separate PR for the refactors to workflow assignment.

srallen commented on app/redux/ducks/classify.js:
```diff
@@ -346,6 +363,56 @@ export function saveAnnotations(annotations) {
+    return Promise.all([awaitWorkflow(workflowId), awaitTranslation])
+    .then(([workflow]) => {
+      if (!workflow) {
+        const error = new Error(`workflow ${workflowId}: empty repsonse from Panoptes.`);
```
👍

eatyourgreens replied:
I misspelt response!

eatyourgreens commented:
I don’t think there’s any big hurry on this. The bug fix is for a bug that inconveniences gold standard classifier, but doesn’t affect classification quality. The rest of the changes are tidying up existing code without adding any new features. 

I think it can wait until the new year.

eatyourgreens commented:
On second thoughts, I've removed the fake 404 error for undefined responses from the workflow request. That promise can't resolve if the workflow ID doesn't exist in Panoptes, so the workflow ID is valid and clearing preferences would lead to bugs like #5173. Empty responses raise an error, but don't reset user preferences.

srallen commented:
@eatyourgreens I think line 389 could change to be less specific. I'm not sure if it is a good idea to fake http status codes. Maybe only run the code to clear the preferences if there isn't an error status code or if the status code is 400 and greater?

eatyourgreens commented:
Preferences should not be cleared if there isn’t a status, since that gives us bugs like #5173. If there’s no status, the bug could be in the API client or in PFE itself and we’d be resetting workflow promotion for valid workflows. 

A more generic handler is a good idea, to catch permissions errors. In that case, the 404 handler in the Workflow Selection component should also be changed, to keep behaviour consistent. Maybe something to consider outside this PR?

eatyourgreens commented:
I've added tests to spec error handling for various error cases. I also simplified the promise chain and moved the error for undefined workflows so that it's with the API request in the code.

srallen commented on app/redux/ducks/classify.spec.js:
```diff
@@ -576,5 +576,164 @@ describe('Classifier actions', function () {
+    describe('on error', function () {
+      describe('404 API errors', function () {
+        before(function () {
+          apiClient.type.restore();
```
Non-blocking comment, but calling this here indicates to me that a test stub isn't being correctly restored elsewhere. Is this why this is here? If so, I think it'd be better to figure out what set of tests that don't have an after test block function defined to handle this. 

I prefer the use `after` or `afterEach` for cleaning up spies and stubs with `resetHistory` or `restore` and using `before` or `beforeEach` to setup for running tests.

eatyourgreens replied:
This is my terrible code. It’s overriding the mock API client that’s set up in the before block, which mocks the client for successful responses.
